src/visitors.rs

```diff
@@ -679,21 +693,65 @@ impl<'a, 'b: 'a, 'tcx: 'b> MirVisitor<'a, 'b, 'tcx> {
                 .constant_value_cache
                 .check_if_std_intrinsics_unreachable_function(&fun)
             {
-                let span = self.current_span;
-                let mut err = self.session.struct_span_warn(
-                    span,
-                    "Control might reach a call to std::intrinsics::unreachable.",
-                );
-                (self.emit_diagnostic)(&mut err, &mut self.buffered_diagnostics);
+                //todo: use theorem prover to make sure that the entry condition is not known to be false.
+                if self
+                    .current_environment
+                    .entry_condition
+                    .as_bool_if_known()
+                    .unwrap_or(false)
+                {
+                    let span = self.current_span;
+                    let mut err = self.session.struct_span_warn(
+                        span,
+                        "Control reaches a call to std::intrinsics::unreachable.",
+                    );
+                    (self.emit_diagnostic)(&mut err, &mut self.buffered_diagnostics);
+                } else {
+                    self.preconditions.push(
+                        self.current_environment
+                            .entry_condition
+                            .not(Some(self.current_span)),
+                    );
+                }
             } else if self
                 .constant_value_cache
                 .check_if_std_panicking_begin_panic_function(&fun)
             {
-                let span = self.current_span;
-                let mut err = self
-                    .session
-                    .struct_span_warn(span, "Execution might panic.");
-                (self.emit_diagnostic)(&mut err, &mut self.buffered_diagnostics);
+                let mut msg = {
+                    if let Expression::CompileTimeConstant(ConstantDomain::Str(ref msg)) =
+                        actual_args[0].domain.expression
+                    {
+                        msg.as_str()
+                    } else {
+                        "Execution might panic."
+                    }
+                };
+                // Since the assertion was conjoined into the entry condition because of the logic
+                // of debug_assert!, and this condition was simplified, we cannot distinguish
+                // between the case of maybe reaching a definitely false assertion from the case of
+                // definitely reaching a maybe false assertion.
+                // We therefore report both cases, even though the first would be a candidate for being
+                // an inferred precondition, rather than a diagnostic.
+                //todo: we need our own library for contract assertions so that we can distinguish
+                // such cases.
+                if msg.starts_with("assertion failed:")
```

Comment:
Wah? Checking string values and replacing them seems weird. I'm too lazy to check to see where this string is set, but don't you control this? Can't you pass something other than a string?

Reply:
I'm afraid not. The string is generated by Rustc. I am not happy about this either, but this is the best I can do at present.